test/source/mock/backend/backend-endpoints.ts

```diff
@@ -53,4 +50,16 @@ const throwIfNotPost = (req: IncomingMessage) => {
   if (!isPost(req)) {
     throw new HttpClientErr('Backend mock calls must use POST method');
   }
+};
+
+const throwIfInvalidOrMissingIdToken = (req: IncomingMessage) => {
```

Comment:
Let's rename it to `getEmailFromIdTokenOrThrow` as `throwIf...` doesn't suppose a return value.

We can merge this, however, I noticed that the API endpoint of `/api/message/upload` is called twice successfully. We can discuss and analyze in a separate issue whether we can retain `UploadedMessageData` from the first successful call to re-use it in the subsequent sending.

Reply:
What do you mean called twice? was this api called twice somewhere?
> I noticed that the API endpoint of /api/message/upload is called twice successfully.

Reply:
> What do you mean called twice? was this api called twice somewhere?

Yes, in `compose - test compose after reconnect account` test, the endpoint `/api/message/upload` was accessed twice -- before account reconnection and after account reconnection, both times successfully (the authentication header was obviously valid) -- so as I understand the message was uploaded to the backend twice (I'm planning to re-check the FES upload too)

Reply:
I checked but it's called twice because of below reason.
When it tries to send pwd-protected message first time(it calls upload api 1st time), it returns error because id_token is missing.
After that app asks user to re-authorize google extension and then re-tries to send message again. (this time it calls upload api 2nd time)

Reply:
> When it tries to send pwd-protected message first time(it calls upload api 1st time), it returns error because id_token is missing.

Sending a PWD-protected message message consists of 2 phases:
1) uploading the PWD-encrypted message to a flowcrypt/FES server, getting message link(s)
2) sending a message containing links/pubkey-encrypted message to `gmail`.
Which phase gives an error? In the mock environment, only the 2. Perhaps, it's worthwhile to update the mock environment/test to mimic the real situation more precisely.

Anyway, this isn't critical, just a consideration.
A possible scenario:
1) PWD-encrypted message gets successfully uploaded, we have `UploadedMessageData`
2) sending message(s) to `gmail` fails (for whatever reason), but auto-retried (e.g. after reconnecting the account)
3) If possible, PWD-encrypted message shouldn't be re-uploaded, provided the inputs are the same (the sender, the recipients, the text etc.) -- which is probably true for auto-resends.
4) sending message(s) to `gmail` using either previous `UploadedMessageData` or the previously built messages -- but the point of concern here is the situation when the user "reconnected" to another gmail account if that's possible.

Reply:
I think 1st phase gives error in mock environment.
1st phase tries to upload PWD-encrypted message to FES using /message/upload url. (Mock environment returns error when id_token is missing or invalid)
Therefore, I think current logic is correct. 
> Which phase gives an error? In the mock environment, only the 2.

Reply:
What I saw during the test was:
1) The mock FES endpoint accepted the token as valid.
2) The mock Google endpoint rejected the token as invalid.
Is this a peculiarity of the test setup or a possible live situation? Please describe your thoughts on this.

Anyway, Google endpoint may possibly reject a message for reasons other than invalid token. The point of #4801 is when we have successfully uploaded the pwd-encrypted message in phase 1, we should retain the message tokens (returned by FES) by reasonably simple means, to re-use them in automatic google-message re-sends. Of course, these tokens should be invalidated if a recipient of the message is added (removed?) or the message text is changed.

So #4801 is worth considering without direct relation to the invalid token issue.

Reply:
Hmm... I didn't experience this. So mock FES treats id_token as valid but mock google endpoint treats id_token as invalid?
Could you tell me reproduce steps?
> What I saw during the test was:

Reply:
> Could you tell me reproduce steps?

The test `compose - test compose after reconnect account`

Reply:
👍  let me check

Reply:
> So mock FES treats id_token as valid

Not FES, `backend` -- flowcrypt API. but this doesn't really matter

Reply:
Just finished checking and I now fully understand your concern.
But if we want to change like above way, we need to change compose send infra and need to change several tens of lines of codes and might make code messy.
Moreover, I think it's fine to call /message/upload twice because id_token will change after google re-auth(or something like that) and backend might need to return different data than previous one.

Let me know your thought @rrrooommmaaa